### DISCUSSION TITLE
T9 Keyboard

### DISCUSSION THREAD

**Discussion by sofia-bg** [UNC-Project-Open-AAC/OS-DPI#242] [Help]
This is a very poor prototype. It only works to change the letter when there is only a single letter present. For example A will change to B and then C but if you press A again it keeps adding A on successive presses. I tried to use filters but was struggling to make them work. For example #label = $label then nothing shows up in the grid? 

Any ideas for improvements? 

[T9v2.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14395166/T9v2.zip)

**Comment by gbishop** [comment]
I don't think you are implementing T9 as described [here](https://en.wikipedia.org/wiki/T9_(predictive_text)). This feels like you intend a multitap approach. 

In T9 for the word "home" you'd type 4663. In a multitap system you'd type 44666,wait,633.

Is your goal a multitap system?

**Reply by sofia-bg** [reply]
yep sorry was using the wrong term, was trying multitap... could also just try real T9

**Reply by sofia-bg** [reply]
but I have been curious about what it would look like for other content to have a multi-tap system, just figured starting with letters to figure out the programming would be a manageable start. I knew too that tackling the wait piece to get a space would be a challenge for me but was hoping to get the multi tap working better first before trying to tackle that

**Reply by gbishop** [reply]
I suggest thinking about how to push the complexity into the content. Suppose your spreadsheet was like this?

| label | letter | previous | row | column |
| ----- | ------ | -------- | --- | ------ |
| aBc   | B      | A        | 1   | 1      |
| abC   | C      | B        | 1   | 1      |
| Abc   | A      | C        | 1   | 1      |
| Abc   | A      | \*       | 1   | 1      |
| dEf   | E      | D        | 2   | 2      |
| deF   | F      | E        | 2   | 2      |
| Def   | D      | F        | 2   | 2      |
| Def   | D      | \*       | 2   | 2      |

I'm using the capitalization to help me keep track of what is happening, it isn't important to the design, though it might be cool to communicate to the user what hitting the key will do. 

I'm thinking there is a state variable $current which records the letter you're currently cycling. It starts empty and gets set to the last letter emitted. The Action has a condition for #previous = '*' which adds the letter, otherwise the letter gets replaced.

Does that help?

**Reply by gbishop** [reply]
You might be able to avoid the extra row with another test based on $current.

**Reply by gbishop** [reply]
Oh, and the Grid is filtering on #previous = $current

**Comment by gbishop** [comment]
Here is a CSV with the 1000 common English words in decreasing frequency encoded into T9. Each letter group is represented by its first letter. The user hits the keys representing the groups then uses another key (encoded here at +) to step through the words with that encoding. For example "call" is AAJJ and "ball" is AAJJ+.

[T9-common-words.csv](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14412656/T9-common-words.csv)

**Reply by gbishop** [reply]
It occurs to me we could represent the letters with SVG images and let the label be the word that is currently predicted. So you hit TUV and the prediction for the GHI key becomes "the", the prediction for MNO is "to", for PQR is "up", for WXYZ is "two", etc. 

It might be worse than not showing them... 

I think it would be straightforward to support both multitap and T9 in the same interface.

**Reply by gbishop** [reply]
Here is a minimal T9 keyboard with 1000 common words. It needs styling for sure. Adding a backspace and delete key would be good. 

[T9.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14434789/T9.zip)

**Reply by sofia-bg** [reply]
Thanks - I'll take a look later today

**Reply by gbishop** [reply]
Here is a design that does T9 and multitap.
[T9MT.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14442031/T9MT.zip)
